providers/in-process-flagd/provider.go

```diff
@@ -0,0 +1,644 @@
+		providerConfiguration: &ProviderConfiguration{},
+		isReady:               make(chan struct{}),
+		connectionInfo: connectionInfo{
+			state:           stateNotReady,
+			retries:         0,
+			maxSyncRetries:  defaultMaxSyncRetries,
+			backoffDuration: defaultBackoffDuration,
+		},
+		ofEventChannel: make(chan of.Event),
+		logger:         logger.NewLogger(nil, false),
+	}
+	provider.applyDefaults()   // defaults have the lowest priority
+	FromEnv()(provider)        // env variables have higher priority than defaults
+	for _, opt := range opts { // explicitly declared options have the highest priority
+		opt(provider)
+	}
+
+	if provider.providerConfiguration.SourceConfig.URI == "" {
+		log.Fatal(errors.New("no sync source configuration provided"))
+	}
+
+	s := store.NewFlags()
+
+	s.FlagSources = append(s.FlagSources, provider.providerConfiguration.SourceConfig.URI)
+	s.SourceMetadata[provider.providerConfiguration.SourceConfig.URI] = store.SourceDetails{
+		Source:   provider.providerConfiguration.SourceConfig.URI,
+		Selector: provider.providerConfiguration.SourceConfig.Selector,
+	}
+
+	var err error
+	provider.syncSource, err = syncProviderFromConfig(provider.logger, provider.providerConfiguration.SourceConfig)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	provider.evaluator = eval.NewJSONEvaluator(
+		provider.logger,
+		s,
+		eval.WithEvaluator(
+			"fractionalEvaluation",
+			eval.NewFractionalEvaluator(provider.logger).FractionalEvaluation,
+		),
+		eval.WithEvaluator(
+			"starts_with",
+			eval.NewStringComparisonEvaluator(provider.logger).StartsWithEvaluation,
+		),
+		eval.WithEvaluator(
+			"ends_with",
+			eval.NewStringComparisonEvaluator(provider.logger).EndsWithEvaluation,
+		),
+		eval.WithEvaluator(
+			"sem_ver",
+			eval.NewSemVerComparisonEvaluator(provider.logger).SemVerEvaluation,
+		),
+	)
+
+	dataSync := make(chan sync.DataSync, 1)
+
+	go provider.watchForUpdates(dataSync)
+
+	if err := provider.syncSource.Init(provider.ctx); err != nil {
+		log.Fatal("sync provider Init returned error: %w", err)
+	}
+
+	// Start sync provider
+	go provider.startSyncSources(dataSync)
+
+	return provider
+}
+
+func (p *Provider) startSyncSources(dataSync chan sync.DataSync) {
+	if err := p.syncSource.Sync(p.ctx, dataSync); err != nil {
+		p.handleConnectionErr(
+			fmt.Errorf("error during source sync: %w", err),
+			func() {
+				p.startSyncSources(dataSync)
+			},
+		)
+	}
+}
+
+func (p *Provider) watchForUpdates(dataSync chan sync.DataSync) error {
+	for {
+		select {
+		case data := <-dataSync:
+			// resync events are triggered when a delete occurs during flag merges in the store
+			// resync events may trigger further resync events, however for a flag to be deleted from the store
+			// its source must match, preventing the opportunity for resync events to snowball
+			if resyncRequired := p.updateWithNotify(data); resyncRequired {
+				go func() {
+					p.tryReSync(dataSync)
+				}()
+			}
+			if data.Type == sync.ALL {
```

Comment:
As of right now, we don't expect any messages other than `sync.ALL`. We may want to log some debug or warning messages if we get others - cc @Kavindu-Dodan

Reply:
yeah, +1 to log other states.